src/libsystemd/sd-netlink/sd-netlink.c

```diff
@@ -705,32 +705,30 @@ static int time_callback(sd_event_source *s, uint64_t usec, void *userdata) {
 
 static int prepare_callback(sd_event_source *s, void *userdata) {
         sd_netlink *nl = userdata;
-        int r, e;
+        int r, enabled;
         usec_t until;
 
         assert(s);
         assert(nl);
 
-        e = sd_netlink_get_events(nl);
-        if (e < 0)
-                return e;
-
-        r = sd_event_source_set_io_events(nl->io_event_source, e);
+        r = sd_netlink_get_events(nl);
         if (r < 0)
                 return r;
 
-        r = sd_netlink_get_timeout(nl, &until);
+        r = sd_event_source_set_io_events(nl->io_event_source, r);
         if (r < 0)
                 return r;
-        if (r > 0) {
-                int j;
 
-                j = sd_event_source_set_time(nl->time_event_source, until);
-                if (j < 0)
-                        return j;
+        enabled = sd_netlink_get_timeout(nl, &until);
+        if (enabled < 0)
+                return enabled;
+        if (enabled > 0) {
+                r = sd_event_source_set_time(nl->time_event_source, until);
+                if (r < 0)
+                        return r;
         }
 
-        r = sd_event_source_set_enabled(nl->time_event_source, r > 0);
+        r = sd_event_source_set_enabled(nl->time_event_source, enabled > 0);
```

Comment:
hmm, actually this would set the time event to SD_EVENT_ON, but I am pretty sure for time events SD_EVENT_ONESHOT is more appropriate